R/functions-util.R

```diff
@@ -70,3 +70,22 @@ setAs("logical", "factor", function(from, to) factor(from))
 sanitize_file_name <- function(x) {
     file.path(normalizePath(dirname(x)), path_sanitize(basename(x)))
 }
+
+#' Helper function that matches `x` against `mz` (using the `closest` function)
+#' and returns the indices of `x` that match any of the values in `mz`. The
+#' function takes care of sorting `x` and `mz` and deals also with missing
+#' values.
+#'
+#' @return `integer` with the indices of values in `x` that are not `NA` and
+#'     are matching any of the values in `mz` given `ppm` and `tolerance`.
+#'
+#' @noRd
+#'
+#' @author Johannes Rainer
+.values_match_mz <- function(x, mz, ppm = 20, tolerance = 0) {
+    keep <- which(!is.na(x))
+    idx <- order(x[keep])
+    mtch <- closest(x[keep][idx], sort(mz), tolerance = tolerance, ppm = ppm,
+                    duplicates = "keep", .check = FALSE)
+    keep[!is.na(mtch[order(idx)])]
```

Comment:
The `order` and `idx` was important (otherwise I wouldn't do such *costly* operations) - but I have to check it again. Thanks for the `common` suggestion - I'll have a try with that and see how it works/compares.

Reply:
And in fact your implementation is still (slightly) faster:

```r
> pmz <- c(12.4, 15, 3, 12.4, 3, 1234, 23, 5, 12.4, NA, 3)
> mz <- c(200, 12.4, 3)
> library(microbenchmark)
> microbenchmark(.values_match_mz(pmz, mz), .values_match_mz2(pmz, mz))
Unit: microseconds
                       expr     min       lq     mean   median       uq     max
  .values_match_mz(pmz, mz) 235.827 272.2785 292.6092 284.0745 317.7635 416.735
 .values_match_mz2(pmz, mz) 178.463 205.5580 221.5553 212.7210 224.7145 446.566
 neval cld
   100   b
   100  a 
> pmz <- rep(pmz, 100)
> mz <- rep(mz, 100)
> microbenchmark(.values_match_mz(pmz, mz), .values_match_mz2(pmz, mz))
Unit: microseconds
                       expr     min       lq     mean   median       uq     max
  .values_match_mz(pmz, mz) 318.139 349.4490 394.0592 376.8095 417.4760 704.230
 .values_match_mz2(pmz, mz) 267.281 298.0505 330.7795 320.6225 349.9465 572.168
 neval cld
   100   b
   100  a 

```

Reply:
I've now changed it to your implementation.

Reply:
Unfortunately I don't have an idea how to remove the last `sort` step but you could avoid the `which(!is.na(x))` call and get minimal more speed:
``` r
library("microbenchmark")
library("MsCoreUtils")
#> 
#> Attaching package: 'MsCoreUtils'
#> The following object is masked from 'package:stats':
#> 
#>     smooth

pmz <- c(12.4, 15, 3, 12.4, 3, 1234, 23, 5, 12.4, NA, 3)
mz <- c(200, 12.4, 3)

.values_match_mz2 <- function(x, mz, ppm = 20, tolerance = 0) {
    keep <- which(!is.na(x))
    idx <- order(x[keep])
    cmn <- common(x[keep][idx], sort(mz), tolerance = tolerance, ppm = ppm,
                  duplicates = "keep", .check = FALSE)
    sort(keep[idx][cmn])
}

.values_match_mz3 <- function(x, mz, ppm = 20, tolerance = 0) {
    o <- order(x, na.last = NA) # na.last = NA will remove NA
    cmn <- common(x[o], sort(mz), tolerance = tolerance, ppm = ppm,
                  duplicates = "keep", .check = FALSE)
    sort(o[cmn])
}
microbenchmark(.values_match_mz2(pmz, mz), .values_match_mz3(pmz, mz), check = "identical")
#> Unit: microseconds
#>                        expr     min      lq     mean   median       uq
#>  .values_match_mz2(pmz, mz) 114.598 117.081 185.6789 118.6825 122.5355
#>  .values_match_mz3(pmz, mz) 111.121 113.789 324.7867 114.7715 120.4950
#>        max neval
#>   6300.412   100
#>  20524.152   100
```

<sup>Created on 2022-01-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Thanks! I learned again something new :)